### PR TITLE
2 script to convert sr14 land use files to sr15

### PR DESCRIPTION
**Pull request by ednaaguilar**
Adding script to convert Series 14 land use data to Series 15 data along with read me.

### CODE REVIEW

File: conversion_sr14_sr15/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+|ABM3 land use| Source of SR15 MGRAs, TAZs, acres, park space, parking costs, and a few other fields held constant in converted MGRA data.|
+|MGRA SR14 to SR15 crosswalk| Crosswalk used for conversion. Created via a separate process in which the centroid of each SR14 MGRA was geocoded to the polygon of SR15 MGRAs.|
+
+## 2.2 Data Processing
+### Households and Persons 
+The synthetic household and person files undergo minimal processing. For the household file, the MGRA column is updated using the SR14 to SR15 crosswalk and the column “hworkers” is renamed to “num_workers”. For the persons file, column ‘miltary’ is renamed to ‘military’, and columns ‘indcen’, ‘weeks’, ‘hours’, ‘rac1p’, ‘hisp’, and ‘version’ are removed, as they are not used in the modeling process. These files are used to calculate household and person totals in the converted SR15 land use file. 
+
+### Land Use
+The land use file undergoes extensive processing. There are three major procedures implemented depending on the columns: 
+1) The converted SR14 synthetic population is used to populate household and person total columns
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/README.md]: Need clarification: the converted SR14 synthetic population is SR15 files, correct?

File: conversion_sr14_sr15/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+|hs, hs_sf, hs_mf, hs_mh, enrollgradekto8, enrollgrade9to12, eollegeenroll, othercollegeenroll, hotelroomtotal|Obtained by applying MGRA crosswalk and summing together rows in the same MGRA |
+
+
+## 2.3 Consistency Checks
+Several checks were conducted to guarantee consistency across the converted synthetic population and land use files. Specifically, we checked the following in the converted SR15 land use file:
+-	Same number of MGRAs as the ABM3 land use file
+-	Taz, luz_id, and other fields were unchanged from the ABM3 land use file
+-	Sum of pop = total population records in synthetic population
+-	Sum of hh = total household records in synthetic population
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/README.md:50]: I checked the current MGRA based land use and synthetic population files and found the total population and household population are not equal to each other. Why do we need to have the checks  (line 47 & 48)?

File: conversion_sr14_sr15/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    - Series 14 synthetic person file
+    - Series 14 synthetic household file
+    - Series 14 land use file
+    - ABM3 land use file
+    - SR14 to SR15 MGRA crosswalk 
+2.	Open the config.yaml file and update the input and output directories as well as the input and output file names. 
+
+    ![image](images/config.png)
+
+3.	Open a terminal and navigate to the folder where conversion_sr14_sr15.py is saved.
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/README.md]: any specific environment needed?

File: conversion_sr14_sr15/config.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Input settings
+input:
+  input_dir: '/home/eamcevoy/edna.aguilar/OneDrive - Resource Systems Group, Inc/Documents/20_Projects/SANDAG/inputs'
+  filenames:
+    households: 'households.csv'
+    persons: 'persons.csv'
+    mgra_xwalk: 'xref_mgra13-15.csv'
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/config.yaml:8]: Is this crosswalk file same format as the one that I shared before? If not, could you share it?

File: conversion_sr14_sr15/conversion_sr14_sr15.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    def run(self):
+        hh_s15 = self.converter.convert_hh()
+        per_s15 = self.converter.convert_per()
+        landuse_s15 = self.converter.convert_landuse(hh_s15, per_s15)
+
+        os.chdir(self.config_loader.config['output']['output_dir'])
+        hh_s15.to_csv(self.config_loader.config['output']['filenames']['households'], index=False)
+        per_s15.to_csv(self.config_loader.config['output']['filenames']['persons'], index=False)
+        landuse_s15.to_csv(self.config_loader.config['output']['filenames']['land_use'], index=False)
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/conversion_sr14_sr15.py:145]: should the land use output file be "mgra15_based_input[year].csv"?

File: conversion_sr14_sr15/config.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    land_use: 'mgra13_based_input2035.csv'
+    land_use_abm3: 'mgra15_based_input2035.csv'
+
+# Output settings
+output:
+  output_dir: '/home/eamcevoy/edna.aguilar/OneDrive - Resource Systems Group, Inc/Documents/20_Projects/SANDAG/outputs'
+  filenames:
+    households: 'households.csv'
+    persons: 'persons.csv'
+    land_use: 'mgra13_based_input2035.csv'
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/config.yaml]: should the output land_use be "mgra15_based_input2035.csv"?

File: conversion_sr14_sr15/images/config.png
Comment (JiaXu1024) [inline conversion_sr14_sr15/images/config.png:1]: The land use output should be "mgra15_based_input2035.csv", please update the pic.

**Review by JiaXu1024 — CHANGES_REQUESTED** [review]
I have added some comments, please address before merging

File: conversion_sr14_sr15/README.md (hunk 1 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/README.md]: Correct, the language has been updated to say: "The "SR15" synthetic population..."

File: conversion_sr14_sr15/README.md (hunk 3 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/README.md]: No special environment is required so long as the base anaconda3 environment is on a machine. Text has been added to line 58 in the Readme to clarify.

File: conversion_sr14_sr15/config.yaml (hunk 4 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/config.yaml:8]: It is the same format as the one you provided

File: conversion_sr14_sr15/config.yaml (hunk 6 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/config.yaml]: Yes, it should be mgra15, the file has been updated.

File: conversion_sr14_sr15/images/config.png
Comment (ednaaguilar) [reply conversion_sr14_sr15/images/config.png:1]: Picture was updated.

File: conversion_sr14_sr15/conversion_sr14_sr15.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    def convert_hh(self):
+        hh_s14 = self.data_loader.hh_s14
+        hh_s14['mgra'] = hh_s14['mgra'].replace(self.mgra_xwalk_dict)
+        hh_s15_converted = hh_s14.rename(columns={'hworkers': 'num_workers'})
+        return hh_s15_converted
+
+    def convert_per(self):
+        per_s14 = self.data_loader.per_s14
+        per_s14.rename(columns={'miltary': 'military'}, inplace=True)
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/conversion_sr14_sr15.py]: SR15 persons file in ABM3 has the column 'miltary', not "military"

File: conversion_sr14_sr15/conversion_sr14_sr15.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        per_s15_converted = per_s14.drop(columns=['indcen', 'weeks', 'hours', 'rac1p', 'hisp', 'version'])
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/conversion_sr14_sr15.py]: SR15 persons file contains the columns "indcen,weeks,hours,rac1p,hisp,version,naics2_original_code,soc2", please add those. @ednaaguilar 
My test run failed due to the missing columns:
image

Comment (ednaaguilar) [reply conversion_sr14_sr15/conversion_sr14_sr15.py]: I updated the script to keep all the columns from SR14 persons file in the SR15 persons file. The log file mentions that naics2_original_code and soc2 are missing but they should be in the converted persons file.

File: conversion_sr14_sr15/conversion_sr14_sr15.py (hunk 7 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/conversion_sr14_sr15.py]: This has been updated

File: conversion_sr14_sr15/config.yaml (hunk 4 above)
Comment (JiaXu1024) [reply conversion_sr14_sr15/config.yaml:8]: How about crosswalk for TAZ? The households file has the column TAZ, and it was not converted from SR14 to SR15.

File: conversion_sr14_sr15/conversion_sr14_sr15.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        self.landuse_abm3 = pd.read_csv(self.config['input']['filenames']['land_use_abm3'])
+
+class Converter:
+    def __init__(self, data_loader):
+        self.data_loader = data_loader
+        self.mgra_xwalk_dict = self.data_loader.mgra_xwalk.set_index('MGRA13')['MGRA15'].to_dict()
+
+    def convert_hh(self):
+        hh_s14 = self.data_loader.hh_s14
+        hh_s14['mgra'] = hh_s14['mgra'].replace(self.mgra_xwalk_dict)
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/conversion_sr14_sr15.py:48]: How about column "taz"? Should it also be converted from SR14 to SR15 using a crosswalk file?

File: conversion_sr14_sr15/conversion_sr14_sr15.py (hunk 8 above)
Comment (JiaXu1024) [reply conversion_sr14_sr15/conversion_sr14_sr15.py]: > I updated the script to keep all the columns from SR14 persons file in the SR15 persons file. The log file mentions that naics2_original_code and soc2 are missing but they should be in the converted persons file.

I believe the columns "naics2_original_code" and "soc2" are new in SR15, not available in SR14. Could you add those in the converted persons file?

File: conversion_sr14_sr15/README.md (hunk 2 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/README.md:50]: We want to make the files internally consistent

File: conversion_sr14_sr15/conversion_sr14_sr15.py (hunk 9 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/conversion_sr14_sr15.py:48]: TAZ column has been converted to SR15

File: conversion_sr14_sr15/conversion_sr14_sr15.py (hunk 8 above)
Comment (ednaaguilar) [reply conversion_sr14_sr15/conversion_sr14_sr15.py]: These two columns have been added by drawing a random sample from an existing SR 15 synthetic person file. A SR 15 synthetic person file is now a required input file.

File: conversion_sr14_sr15/README.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    - Series 14 land use file
+    - ABM3 land use file
+    - SR14 to SR15 MGRA crosswalk 
+    - SR14 to SR15 TAZ crosswalk
+    - Series 15 synthetic person file
+2.	Open the config.yaml file and update the input and output directories as well as the input and output file names. 
+
+    ![image](images/config.png)
+
+3.	Open a terminal and navigate to the folder where conversion_sr14_sr15.py is saved.
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/README.md]: Should be "open an Anaconda Prompt (Anaconda 3) ..." @ednaaguilar

File: conversion_sr14_sr15/README.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+|SR14 land use file| Source of employment and income to be used in converted MGRA data file.|
+|ABM3 land use| Source of SR15 MGRAs, TAZs, acres, park space, parking costs, and a few other fields held constant in converted MGRA data.|
+|MGRA SR14 to SR15 crosswalk| Crosswalk used for conversion. Created via a separate process in which the centroid of each SR14 MGRA was geocoded to the polygon of SR15 MGRAs.|
+|TAZ SR14 to SR15 crosswalk| Crosswalk used for conversion.|
+|SR15 synthetic person file| Generates naics2_original_code and soc2 columns for SR15 synthetic person file by drawing a uniform random sample.|
+
+## 2.2 Data Processing
+### Households and Persons 
+The synthetic household and person files undergo minimal processing. For the household file, the MGRA and TAZ columns are updated using the SR14 to SR15 crosswalks, and the column “hworkers” is renamed to “num_workers”. For the person file, "naics2_original_code" and "soc2" are generated by drawing a uniform random sample from an existing SR15 synthetic person file. The converted SR15 person and household files are used to calculate household and person totals in the converted SR15 land use file.
```
Comment (JiaXu1024) [inline conversion_sr14_sr15/README.md:21]: For the method of generating 'naics2_original_code' and 'soc2', does the uniform random sample change every time while the tool is running? I wonder if the outputs are different each time with the same input files? @ednaaguilar

Comment (JiaXu1024) [reply conversion_sr14_sr15/README.md:21]: And just quick check the output 'persons' file, and I found the  'naics2_original_code' and 'soc2' have 0 values with workers (employ = [1,2]). I don't think it's correct. Could you check your code and make an explanation? @ednaaguilar

**Pull request merged**